File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
   public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int cd= Integer.MAX_VALUE;
        int cn= target;
        for (int i = 0; i < nums.length - 2; i++) {
            int l= i + 1;
            int h= nums.length-1;
            while(l<h-1+1) {
                int s= nums[i] + nums[l]+nums[h];
                if (s==target) {
                    return s;
                }else if (s< target) {
                    l++;
                }else {
                    h--;
                }
                int d = Math.abs(target-s);
                if (cd>d) {
                    cd= d;
                    cn= s;
                }
            }
        }
        return cn;
    }
}

File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        # initialize variables
        solution = sum(nums[:3])
        # sort the array
        nums.sort()
        # use three pointers one for each element other for two 
        for number in range(len(nums)-2):
            left = number+1
            right = len(nums)-1
            current_sum = 0
            while left < right:
                current_sum =  nums[left]+nums[right]+nums[number]
                if current_sum < target:
                    left += 1
                elif current_sum > target:
                    right -= 1
                else:
                    return current_sum
                if (abs(current_sum-target) < abs(solution-target)):
                    solution = current_sum
        return solution
    